src/Tuura/Concept/FSM/Translation.hs

```diff
@@ -81,46 +81,55 @@ translateFSM circuitName ctype signs = do
+              invariantNos = map encToInt invariants
               reachableArcs = removeUnreachables allArcs reachables
               inputSigns = filter ((==Input) . interface circuit) signs
               outputSigns = filter ((==Output) . interface circuit) signs
               internalSigns = filter ((==Internal) . interface circuit) signs
-              reachableInvariants = filter (`elem` reachables) invariants
-              unreachables = ([0..2^(length signs) - 1] \\ invariants) \\ reachables
-          if (reachableInvariants /= [])
-              then reachableInvariantStateError reachableInvariants
-            else do
-              let reachReport = genReachReport unreachables
-              genFSM inputSigns outputSigns internalSigns (map show reachableArcs) (show initialState) reachReport
-      Invalid errs ->
-          "Error. \n" ++ addErrors errs
+              reachableInvariants = filter (\i -> (encToInt i) `elem` reachables) invariants
+              unreachables = ([0..2^(length signs) - 1] \\ invariantNos) \\ reachables
+          case validateFSM signs circuit reachableInvariants (invariant circuit) of
+              Valid -> do
+                let reachReport = genReachReport unreachables
+                genFSM inputSigns outputSigns internalSigns (map show reachableArcs) (show initialState) reachReport
+              Invalid errs -> addErrors errs
+
+      Invalid errs -> addErrors errs
 
 getInitialState :: CircuitConcept a -> [a] -> Int
 getInitialState circuit signs = encToInt state
   where
-    state = map (fromBool  . getDefined . initial circuit ) signs
+    state = map (fromBool  . getDefined . initial circuit) signs
 
 fromBool :: Bool -> Tristate
 fromBool x = if x then triTrue else triFalse
 
-addConsistency :: Ord a => [([Transition a], Transition a)] -> [a] -> [([Transition a], Transition a)]
-addConsistency allArcs signs = nubOrd (allArcs ++ concatMap (\s -> [([rise s], fall s), ([fall s], rise s)]) signs)
+addConsistency :: Ord a => [Causality (Transition a)] -> [a] -> [Causality (Transition a)]
+addConsistency allArcs signs = nubOrd (allArcs ++ concatMap (\s -> [AndCausality (rise s, fall s), AndCausality (fall s, rise s)]) signs)
 
 handleArcs :: [([Transition a], Transition a)] -> [([Transition a], Transition a)]
-handleArcs arcLists = result
+handleArcs xs = map (\m -> (m, effect)) transCauses
         where
-            effect = snd (head arcLists)
-            effectCauses = map fst arcLists
+            effect = snd (head xs)
+            effectCauses = map fst xs
             transCauses = cartesianProduct effectCauses
-            result = map (\m -> (m, effect)) transCauses
+
+validateFSM :: Ord a => [a] -> CircuitConcept a -> [[Tristate]] -> [Invariant (Transition a)] -> ValidationResult a
+validateFSM signs circuit reachInvs invs
+    | invVio == [] && unused == [] = Valid
+    | otherwise = Invalid ((map UndefinedInitialState unused) ++ (map InvariantViolated invVio))
```

Comment:
No need for brackets around `map`. Function application (space) has highest precedence in Haskell.